server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.util.timer;
+
+import org.apache.kafka.common.utils.KafkaThread;
+import org.apache.kafka.common.utils.Time;
+
+import java.util.concurrent.DelayQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class SystemTimer implements Timer {
```

Comment:
I am not sure to understand the benefits of doing this here. My understanding is that `AutoClosable` just adds a `close` method but you still need to class it explicitly or implicitly by using `try-with-resources`. If we want to do this, we would need to bring this change to the `Timer` interface instead of doing it here. I think that this is something that we could consider separately.

Reply:
Many automatic bug finders will flag a warning if an object of class which is marked as AutoCloseable but is not closed properly in the code. That is one advantage of using this interface that I find very useful.

Reply:
Doing is separately is fine for me.